test/functional/api/highlight_spec.lua

```diff
@@ -369,3 +369,218 @@ describe("API: set highlight", function()
+    local ns = get_ns()
+    meths.set_hl(ns, 'Test_hl', highlight1)
+    eq(highlight1, meths.get_hl(ns, { name = 'Test_hl' }))
+  end)
+
+  it('can overwrite attr for cterm', function()
+    local ns = get_ns()
+    meths.set_hl(ns, 'Test_hl', highlight3_config)
+    eq(highlight3_result, meths.get_hl(ns, { name = 'Test_hl' }))
+  end)
+
+  it('only allows one underline attribute #22371', function()
+    local ns = get_ns()
+    meths.set_hl(ns, 'Test_hl', {
+      underdouble = true,
+      underdotted = true,
+      cterm = {
+        underline = true,
+        undercurl = true,
+      },
+    })
+    eq({ undercurl = true, underdotted = true }, meths.get_hl(ns, { name = 'Test_hl' }))
+  end)
+
+  it('can get a highlight in the global namespace', function()
+    meths.set_hl(0, 'Test_hl', highlight2)
+    eq(highlight2, meths.get_hl(0, { name = 'Test_hl' }))
+
+    meths.set_hl(0, 'Test_hl', { background = highlight_color.bg })
+    eq({
+      bg = 12970,
+    }, meths.get_hl(0, { name = 'Test_hl' }))
+
+    meths.set_hl(0, 'Test_hl2', highlight3_config)
+    eq(highlight3_result, meths.get_hl(0, { name = 'Test_hl2' }))
+
+    -- Colors are stored with the name they are defined, but
+    -- with canonical casing
+    meths.set_hl(0, 'Test_hl3', { bg = 'reD', fg = 'bLue' })
+    eq({
+      bg = 16711680,
+      fg = 255,
+    }, meths.get_hl(0, { name = 'Test_hl3' }))
+  end)
+
+  local expected_rgb = {
+    altfont = true,
+    bg = 16776960,
+    bold = true,
+    ctermbg = 10,
+    fg = 16711680,
+    italic = true,
+    nocombine = true,
+    reverse = true,
+    sp = 255,
+    strikethrough = true,
+    underline = true,
+  }
+  local expected = {
+    bg = 16776960,
+    bold = true,
+    ctermbg = 10,
+    fg = 16711680,
+    sp = 255,
+    underline = true,
+  }
+  local expected_undercurl = {
+    bg = 16776960,
+    ctermbg = 10,
+    fg = 16711680,
+    sp = 255,
+    undercurl = true,
+    underline = true,
+  }
+
+  it('nvim_get_hl by id', function()
+    local hl_id = meths.get_hl_id_by_name('NewHighlight')
+
+    command(
+      'hi NewHighlight cterm=underline ctermbg=green guifg=red guibg=yellow guisp=blue gui=bold'
+    )
+    eq(expected, meths.get_hl(0, { id = hl_id }))
+
+    -- Test 0 argument
+    eq('Highlight id out of bounds', pcall_err(meths.get_hl, 0, { id = 0 }))
+
+    eq(
+      'Invalid highlight id: expected Integer, got String',
+      pcall_err(meths.get_hl, 0, { id = 'Test_set_hl' })
+    )
+
+    -- Test all highlight properties.
+    command('hi NewHighlight gui=underline,bold,italic,reverse,strikethrough,altfont,nocombine')
+    eq(expected_rgb, meths.get_hl(0, { id = hl_id }))
+
+    -- Test undercurl
+    command('hi NewHighlight gui=undercurl')
+    eq(expected_undercurl, meths.get_hl(0, { id = hl_id }))
+  end)
+
+  before_each(function()
+    clear()
+    command('hi String guifg=#a6e3a1')
+    command('hi link @string string')
+    command('hi link @string.cpp @string')
+  end)
```

Comment:
If this is only intended for the next test, would be clearer to put the command() calls directly in the test. The `before_each` applies to the entire scope of its parent `describe()`. I _think_ busted will apply this before_each to all of the it() calls above, thus the clear() happens twice because of the other before_each (line 431).

Better to have one before_each near the top, to avoid surprises.